irobotclient/main.py

```diff
@@ -0,0 +1,100 @@
+"""
+Copyright (c) 2017 Genome Research Ltd.
+
+This program is free software: you can redistribute it and/or modify it
+under the terms of the GNU General Public License as published by the
+Free Software Foundation, either version 3 of the License, or (at your
+option) any later version.
+
+This program is distributed in the hope that it will be useful, but
+WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
+Public License for more details.
+
+You should have received a copy of the GNU General Public License along
+with this program. If not, see <http://www.gnu.org/licenses/>.
+"""
+import errno
+
+from os import path
+from requests import Response
+
+from irobotclient import configuration_handler
+from irobotclient import request_formatter
+from irobotclient.custom_exceptions import IrobotClientException
+from irobotclient.request_handler import Requester
+
+
+def _print_error_details(error: OSError):
+    """
+    Print the output of any errors raised.
+
+    Utilises the OSError exception class which the iRobot Client uses for custom exceptions too.
+
+    :param error:
+    :return:
+    """
+    print("Inside _output_error_details") # Beth - debug
+    print(error)
+    exit(error.errno)
+
+
+def _download_data(response: Response, output_dir:str):
+    """
+
+    :param response:
+    :param output_dir:
+    :return:
+    """
+    try:
+        file_name = (path.split(response.url))[1]
+
+        if response.headers["content-type"] == "application/octet-stream":
+            with open(f"{output_dir}{file_name}", "wb") as file:
+                file.write(bytes(response.content))
+        else:
+            with open(f"{output_dir}{file_name}", "w") as file:
+                file.write(response.content)
+    except OSError as err:
+        _print_error_details(err)
+    except:
+        raise IrobotClientException(errno=errno.ECONNABORTED, message="Cannot write content to file.")
+
+
+def _run():
+    try:
+
+        # Set configurations from command line and/or environment.
+        config_details = configuration_handler.run()
+
+        file_extensions = request_formatter.get_file_extensions(config_details.input_file, config_details.no_index)
+        url = request_formatter.get_url_request_path(config_details.url, config_details.input_file)
+        headers = request_formatter.get_header(config_details.token)
+
+        if not file_extensions:
```

Comment:
You're repeating yourself here. If `file_extensions` is empty, then you call the same code as you would in a loop when `file_extensions` isn't empty. You can generalise this to the looping structure by changing `get_file_extensions` to always return a collection containing at least one item (i.e., the originally requested file's extension), plus others (i.e., the indices) if it deems necessary.

What happens when you want to download a file that doesn't have an extension? The extension is a purely mnemonic device, so there's no reason to assume it will always be there.